pkg/querier/blocks_store_querable_streaming.go

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querier
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storegateway/storegatewaypb"
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+	"github.com/grafana/mimir/pkg/util/limiter"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+// StreamingSeries holds the labels of the streaming series and the source to get the chunks
+// for the series.
+type StreamingSeries struct {
+	Labels []mimirpb.LabelAdapter
+	Source StreamingSeriesSource
+}
+
+// StreamingSeriesSource holds the relationship between a stream of chunks from a SeriesChunksStreamReader
+// and the expected position of a series' chunks in that stream.
+type StreamingSeriesSource struct {
+	StreamReader *SeriesChunksStreamReader
+	SeriesIndex  uint64
+}
+
+// SeriesChunksStreamReader is responsible for managing the streaming of chunks from a storegateway and buffering
+// chunks in memory until they are consumed by the PromQL engine.
+type SeriesChunksStreamReader struct {
+	client              storegatewaypb.StoreGateway_SeriesClient
+	expectedSeriesCount int
+	queryLimiter        *limiter.QueryLimiter
+	stats               *stats.Stats
+	log                 log.Logger
+
+	seriesCunksChan chan *storepb.StreamSeriesChunksBatch
+	chunksBatch     []*storepb.StreamSeriesChunks
+	errorChan       chan error
+}
+
+func NewSeriesChunksStreamReader(client storegatewaypb.StoreGateway_SeriesClient, expectedSeriesCount int, queryLimiter *limiter.QueryLimiter, stats *stats.Stats, log log.Logger) *SeriesChunksStreamReader {
+	return &SeriesChunksStreamReader{
+		client:              client,
+		expectedSeriesCount: expectedSeriesCount,
+		queryLimiter:        queryLimiter,
+		stats:               stats,
+		log:                 log,
+	}
+}
+
+// Close cleans up all resources associated with this SeriesChunksStreamReader.
+// This method should only be called if StartBuffering is not called.
+func (s *SeriesChunksStreamReader) Close() {
+	if err := s.client.CloseSend(); err != nil {
+		level.Warn(s.log).Log("msg", "closing storegateway client stream failed", "err", err)
+	}
+}
+
+// StartBuffering begins streaming series' chunks from the storegateway associated with
+// this SeriesChunksStreamReader. Once all series have been consumed with GetChunks, all resources
+// associated with this SeriesChunksStreamReader are cleaned up.
+// If an error occurs while streaming, a subsequent call to GetChunks will return an error.
+// To cancel buffering, cancel the context associated with this SeriesChunksStreamReader's storegatewaypb.StoreGateway_SeriesClient.
+func (s *SeriesChunksStreamReader) StartBuffering() {
+	s.seriesCunksChan = make(chan *storepb.StreamSeriesChunksBatch, 10)
```

Comment:
No science. Having it as 1 was giving goroutine leak errors in unit tests, so I just made it 10. I can make it 2 (just for the test)

Reply:
For streaming from ingesters, we only buffer one batch (plus one being actively read from). What's causing the goroutine leak if we have just one buffered?

Reply:
I think the last message is stuck at sending to the channel so the goroutine does not end. I did not investigate further and channel size of 2 worked. I will make it 1 again and see if we get the error again. Maybe it is hinting some other underlying problem.

Reply:
This was a good call. There was a genuine case of goroutine leak when the call to storegateway was a success, but later querier fails on the block consistency check, leaking the goroutine because we did not close the streamreaders there.